src/visit.ts

```diff
@@ -480,3 +490,29 @@ export function visitEachChild<T extends FunctionlessNode>(
   }
   return assertNever(node);
 }
+
+/**
+ * Like {@link visitEachChild} but it only visits the statements of a block.
+ *
+ * Provides the hoist function that allows hoisting expressions into variable statements above the current statement.
+ */
+export function visitBlock(
+  block: BlockStmt,
+  cb: (stmt: Stmt, hoist: (expr: Expr) => Expr) => Stmt,
+  nameGenerator: DeterministicNameGenerator
+): BlockStmt {
+  return visitEachChild(block, (stmt) => {
+    const nestedTasks: FunctionlessNode[] = [];
+    function hoist(expr: Expr): Expr {
+      const id = new Identifier(nameGenerator.generateOrGet(expr));
+      nestedTasks.push(new VariableStmt(id.name, expr));
+      return id;
+    }
```

Comment:
Why would we do it this way instead of transforming the tree before walking?

Reply:
We are, this loop is running before walking it, it just reshapes the tree before the interpreter runs on it. Its 1:1 with the previous pre-interpreter transform, it just supports arbitrary depth, complexity, and location of integrations.

```ts
for(const i in [func(), func(), func(func())])
```

out

```ts
const 0_expr = func();
const 1_expr = func();
const 2_expr = func();
const 3_expr = func(2_expr);
for(const i in [0_expr, 1_expr, 3_expr])
```

Reply:
I mean why is the hoist running within the generic visit function?

Also, how does it handle `ConditionExpr`?

```ts
for (const i in [cond ? expr_1 : expr_2])
```

Reply:
> Also, how does it handle ConditionExpr?

it should? can add a test. It won't short circuit, but that's a risk of any of the hoisting. We should probably make the actually processing smarter and get rid of the a-normal form at some point, or also inline the conditional statements.

We could also handle specific cases differently like I'll show with async.

But I don't know the difference between

```ts
func(func2())
```

and 

```ts
func(x ? func() : func2())
```

Could also turn turnary statements into if-statements to maintain scoping instead of the a-normal form.

> I mean why is the hoist running within the generic visit function?

Its not running, its an optional callback. When processing a block statement, calls can be hoisted.

Reply:
An attempt to generalize this

https://github.com/functionless/functionless/blob/main/src/asl.ts#L403-L450

Reply:
lol

(step functions)

![image](https://user-images.githubusercontent.com/289213/174500312-8c865989-776b-4d55-b60e-0f4a88afc6b8.png)